File: 3sem/fds/heap_sort.py
def heapify(arr, n, i):
    # Find the largest element among root, left child, and right child
    largest = i  # Initialize largest as root
    left = 2 * i + 1  # Left child index
    right = 2 * i + 2  # Right child index

    # If the left child exists and is greater than the root
    if left < n and arr[left] > arr[largest]:
        largest = left

    # If the right child exists and is greater than the current largest
    if right < n and arr[right] > arr[largest]:
        largest = right

    # If the largest element is not the root
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # Swap
        heapify(arr, n, largest)  # Recursively heapify the affected sub-tree

def heap_sort(arr):
    n = len(arr)

    # Build a maxheap (rearrange the array)
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # Extract elements from the heap one by one
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # Swap
        heapify(arr, i, 0)  # Heapify the root element to maintain the heap property

# Example usage
arr = [12, 11, 13, 5, 6, 7]
heap_sort(arr)
print("Sorted array is:", arr)

File: 4sem/dsal pratical/heapsort.cpp
#include<iostream>
#include<vector>
using namespace std;

void heapify(vector<int> &arr,int n,int i){
    int largest=i;
    int l=2*i+1;
    int r=2*i+2;
    
    if(l<n  && arr[l]>arr[largest])
    {
        largest=l;
    }
    if(r<n  && arr[r]>arr[largest])
    {
        largest=r;
    }
    if(largest!=i){
        swap(arr[i],arr[largest]);
        heapify(arr, n, largest);
        
    }
    
    
}


void heap_sort(vector<int>& arr){
    int n=arr.size();
    for(int i=n/2-1;i>=0;i--){
                heapify(arr, n, i);
    }
    for(int i=n-1;i>=0;i--){
        swap(arr[0],arr[i]);
        heapify(arr,i,0);
    }
}

void printHeap(vector<int> arr){
    for(int i=0;i<arr.size();i++){
        cout<<arr[i]<<" ";
    }
}


int main(){
    
    vector<int> arr = { 9, 4, 3, 8, 10, 2, 5 };
    // Function call
    heap_sort(arr);
    printHeap(arr);
    return 0;
}